src/Firestore/SnapshotTrait.php

```diff
@@ -0,0 +1,155 @@
+<?php
+/**
+ * Copyright 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Firestore;
+
+use Google\Cloud\Core\Timestamp;
+use Google\Cloud\Core\ArrayTrait;
+use Google\Cloud\Core\ValueMapperTrait;
+use Google\Cloud\Core\Exception\NotFoundException;
+
+/**
+ * Methods common to representing Document Snapshots.
+ */
+trait SnapshotTrait
+{
+    use ArrayTrait;
+    use PathTrait;
+
+    /**
+     * Create and return a document snapshot.
+     *
+     * If `$options['data']` is set, no service request will be triggered, and
+     * the returned snapshot will indicate that the document does exist, and
+     * will be populated with the data provided.
+     *
+     * @param DocumentReference $document The parent document.
+     * @param array $options {
+     *     Configuration Options
+     *
+     *     @type bool $exists If set to false, no service request will be
+     *           triggered, and the returned snapshot will indicate that the
+     *           document does not exist. **Defaults to** `false`.
+     *     @type array $data If set, no service request will be triggered, and
+     *           the returned snapshot will indicate that the document does
+     *           exist, and will be populated with the data provided.
+     *     @type string $transactionId The transaction ID to fetch the snapshot.
+     *     @type bool $allowNonExistence If true, a DocumentSnapshot will be
+     *           returned, even if the document does not exist. **Defaults to**
+     *           `true`.
+     *     @type Timestamp $readTime Reads the version of the document at the
+     *           given time. This may not be older than 60 seconds.
+     * }
+     * @return DocumentSnapshot
+     * @throws NotFoundException If the document does not exist, and
+     *         `$options['allowNonExistence'] is `false`.
+     */
+    private function createSnapshot(DocumentReference $document, array $options = [])
```

Comment:
Please take this with a grain of salt since I know virtually nothing about PHP and this Repo: 

It seems weird to me to combine these two very different use cases. Can we split this up into a method that explicitly fetches data from the backend and one that takes the input data as is?